not_tcp/not_tcp.py

```diff
@@ -0,0 +1,100 @@
+"""
+Not TCP is a simple protocol for running HTTP/1.0 sessions over a
+serial stream.
+
+Each packet includes a short header: flags, a session ID, and a body-length.
+
+"""
+
+from amaranth import Module, Signal, unsigned, Const
+from amaranth.lib.wiring import Component, In, Out, Signature
+from amaranth.lib import stream
+from amaranth.lib.data import UnionLayout, ArrayLayout, Struct
+from amaranth.lib.fifo import SyncFIFOBuffered
+import session
+
+
+class Flags(Struct):
+    """
+    Flags in a Not TCP header.
+    """
+
+    # Start-of-stream marker.
+    start: 1
+
+    # End-of-stream marker.
+    end: 1
+
+    # Direction marker: 0 for "client to server", 1 for "server to client"
+    direction: 1
+
+    # Additional bits for the future.
+    unused: 5
+
+
+class Header(Struct):
+    """
+    Layout of a Not TCP header.
+    """
+
+    flags: Flags
+
+    # Session identifier.
+    session: 8
+
+    # Length of the body following the header
+    # (bytes to the next header)
+    length: 8
+
+
+class BusStopSignature(Signature):
+    """
+    Signature of a stop on a Not TCP local bus.
+
+    Attributes
+    ----------
+    upstream:   Stream(8), In
+    downstream: Stream(8), Out
+    """
+
+    def __init__(self):
+        super().__init__({
+            "upstream": In(stream.Signature(8)),
+            "downstream": Out(stream.Signature(8)),
+        })
+
+
+class BusRoot(Component):
+    """
+    Root of a Not TCP bus.
+    Connects the (host) serial lines to the (device-local) bus
+    and sorts packets between upstrema and downstream.
+    """
+
+    bus: Out(BusStopSignature())
+    tx: Out(stream.Signature(8))
+    rx: In(stream.Signature(8))
```

Comment:
Are these the signals going to the StreamStops?

Reply:
The `tx` and `rx` ends are meant to be the USB-serial device. For a proof-of-concept/single-stop, yeah, "just" connect these to upstream (`rx`) and downstream (`tx`).

If we made the `StreamStop` a true ring, where it passes around the unmatched packets, then that would be the same (upstream/downstream; unmatched "!to_host" packets go back to the host to signal "that's an invalid stream").

I was thinking an easier multiplexing might be to mux on stream_id and to the round-robin more centrally. Will play with it, but yeah, this is ~degenerate for now.

In fact, here, let me implement it: eefbcd4

Reply:
👍